### PR TITLE
Detect GA3, GA4 or GTM during installation and suggest migration guides

### PR DESCRIPTION
**Pull request by bx80**
### Description:

Checks for GA3, GA4 or GTM usage on a site when showing the tracking code / getting started screen and if present, will show a link to the appropriate guide on how to migrate to Matomo.

Ref: DEV-3005

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Comment by bx80** [comment]
The installation test was failing because the site content detection values required by the instructions email and tracking code templates were not being populated for fresh installs. I've updated the installation controller to do a site content detection request as part of the install wizard and provide the required values. I've also added a new fixture which will mock the site content detector finding nothing on the site and added this to the installation UI test so that the test doesn't make live requests to the test site.

**Comment by peterhashair** [comment]
@bx80 installation works now, once the tests are passed, will do another review :)

**Review by peterhashair — APPROVED** [review]
Reviewed functionality and code, working as expected. 
The only I am wondering about in the `SiteContentDetector` class, there is a set of test data for the fixture, which works for me, not sure if there is a different approach. 

* [x] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [x] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [x] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [x] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [x] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [x] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [x] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [x] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [x] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [x] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [x] Existing documentation updated if needed

**Comment by bx80** [comment]
@peterhashair If you mean the data in the `getConsentManagerDefinitions()` method, that's not test data, it's the data used to detect the known consent managers and their guide URLs. This PR is branched from #19821, which is where that code is used. We could move the definitions to a set of external files, but I can't see much benefit to this.

**Comment by peterhashair** [comment]
@bx80 sorry I mean, those functions  `private function checkForTestData() : bool`  below that as well.

**Comment by bx80** [comment]
@peterhashair The site content detector gets called as part of the UI tests and while it's possible to mock class methods using PHPUnit for unit tests, I haven't seen a good way to do this with UI tests which will interact with the class method by way of an API call. The nearest existing pattern I found in the codebase was the location provider where a mock location provider class is set as the default provider via an option setting. The approach here is broadly similar, where the test fixture can set test data as an option and the class will check for it and return the test data values if they exist. Not the most elegant solution, but it works and is flexible. I'm open to better ideas though,

@sgiehl Do you have a recommended approach for this scenario?

**Comment by sgiehl** [comment]
> @sgiehl Do you have a recommended approach for this scenario?

I guess we have this solved in various ways. I haven't had a look at the detailed solution here. But I guess I would have tried to somehow solve it using DI. So the class responsible for the detection is loaded using DI. And for ui tests, you could then overwrite the used class with some sort of test class, that returns fake response.

**Comment by bx80** [comment]
Thanks @sgiehl, I've reworked the UI tests to use DI :+1:

**Review by peterhashair — APPROVED** [review]
Looks good to me.

File: plugins/SitesManager/templates/_trackingCodeEmail.twig [hunk 1]
```diff
@@ -2,6 +2,26 @@
...
+{{ 'CoreAdminHome_JSTracking_ConsentManagerEmailNote'|translate }}
+{% endif %}
+
+{% if ga3Used %}
+{{ 'SitesManager_GADetectedEmail'|translate('Google Analytics 3', 'GA', 'https://matomo.org/faq/how-to/migrate-from-google-analytics-3-to-matomo/')|raw }}
+{{ 'CoreAdminHome_JSTracking_ConsentManagerEmailNote'|translate }}
+{% endif %}
+
+{% if ga4Used %}
+{{ 'SitesManager_GA4DetectedEmail'|translate('Google Analytics 4', 'GA', 'https://matomo.org/faq/how-to/migrate-from-google-analytics-4-to-matomo/')|raw }}
```
Comment (AltamashShaikh) [inline plugins/SitesManager/templates/_trackingCodeEmail.twig:16]: @bx80 @sgiehl I cannot seem to find the translation for `GA4DetectedEmail` and `GTMDetectedEmail` in the [translation file](https://github.com/matomo-org/matomo/blob/4.x-dev/plugins/SitesManager/lang/en.json) is this a bug ?

Comment (AltamashShaikh) [reply plugins/SitesManager/templates/_trackingCodeEmail.twig:16]: Its a bug, have fixed it in my upcoming Pr for cloudflare

**Pull request merged**